.github/workflows/PR-build.yml

```diff
@@ -209,44 +207,51 @@ jobs:
       matrix: ${{ fromJson(needs.get-test-cases.outputs.matrix) }}
     steps:
       - name: Check out testing framework
+        if: ${{ needs.changes.outputs.changed == 'true' }}
         uses: actions/checkout@v2
         with:
           repository: 'aws-observability/aws-otel-collector-test-framework'
           path: testing-framework
-        if: ${{ needs.changes.outputs.changed == 'true' }}
 
       - name: Check out Collector
+        if: ${{ needs.changes.outputs.changed == 'true' }}
         uses: actions/checkout@v2
         with:
           path: aws-otel-collector
-        if: ${{ needs.changes.outputs.changed == 'true' }}
 
       - name: Set up JDK 11
+        if: ${{ needs.changes.outputs.changed == 'true' }}
         uses: actions/setup-java@v2
         with:
           distribution: 'zulu'
           java-version: '11'
-        if: ${{ needs.changes.outputs.changed == 'true' }}
 
       - name: Set up terraform
-        uses: hashicorp/setup-terraform@v1.2.1
         if: ${{ needs.changes.outputs.changed == 'true' }}
+        uses: hashicorp/setup-terraform@v1.2.1
 
       - name: restore cached binaries
+        if: ${{ needs.changes.outputs.changed == 'true' }}
         uses: actions/cache@v2
         with:
           key: "cached_binaries_${{ github.run_id }}"
           path: build
-        if: ${{ needs.changes.outputs.changed == 'true' }}
 
       - name: copy binary
-        run: cp -R build aws-otel-collector/build
         if: ${{ needs.changes.outputs.changed == 'true' }}
+        run: cp -R build aws-otel-collector/build
 
       - name: Run test
-        run: | 
-          if [[ -f testing-framework/terraform/testcases/${{ matrix.testcase }}/parameters.tfvars ]] ; then opts="-var-file=../testcases/${{ matrix.testcase }}/parameters.tfvars" ; else opts="" ; fi
-          cd testing-framework/terraform/mock && terraform init && terraform apply -auto-approve -var="testcase=../testcases/${{ matrix.testcase }}" $opts
         if: ${{ needs.changes.outputs.changed == 'true' }}
+        uses: nick-invision/retry@v2
+        with:
+          max_attempts: 3
+          timeout_minutes: 10
+          command: |
+            if [[ -f testing-framework/terraform/testcases/${{ matrix.testcase }}/parameters.tfvars ]] ; then opts="-var-file=../testcases/${{ matrix.testcase }}/parameters.tfvars" ; else opts="" ; fi
+            cd testing-framework/terraform/mock && terraform init && terraform apply -auto-approve -var="testcase=../testcases/${{ matrix.testcase }}" $opts
+          on_retry_command: |
+            cd testing-framework/terraform/mock && terraform destroy -auto-approve
+            sleep $((RANDOM % 10))
```

Comment:
Instead of using sleep, it would be better to use [retry wait seconds](https://github.com/nick-invision/retry#retry_wait_seconds) though

Reply:
I looked at that, but the main purpose of this is to stagger the retries. The hope is that if two tests retry at the same time because they both hit the rate limit, they don't run at exactly the same time again. If I could set a random value to `retry_wait_seconds` then that would work. Not sure if I can.

Reply:
Interesting. So random value for `retry_wait_seconds` is not possible and random % 10 is quite a small range though. Anyway we could use the image from docker hub instead of ECR image to avoid pulling quota?

Reply:
DockerHub has a rate limit as well. I think we're seeing if we can get a rate limit increase for the public ECR. A random 0-10 second delay should be good enough.